File: python/socketd/socketd/transport/server/ServerBase.py
from typing import Callable, Set

from socketd.transport.core.Listener import Listener
from socketd.transport.core.Message import Message
from socketd.transport.core.Processor import Processor
from socketd.transport.core.Session import Session
from socketd.transport.core.listener.SimpleListener import SimpleListener
from socketd.transport.server.Server import Server
from socketd.transport.server.ServerConfig import ServerConfig
from socketd.transport.core.impl.ProcessorDefault import ProcessorDefault
from socketd.transport.core.ChannelAssistant import ChannelAssistant
from socketd.utils.RunUtils import RunUtils


class ServerBase(Server,Listener):
    """
    服务端基类
    """

    def __init__(self, config:ServerConfig, assistant:ChannelAssistant):
        self._config = config
        self._assistant = assistant
        self._isStarted:bool = False

        self._processor: Processor = ProcessorDefault()
        self._sessions:Set[Session] = set()
        self._listener: Listener = SimpleListener()

        self._processor.set_listener(self)

    def get_assistant(self):
        """
        获取通道助理
        """
        return self._assistant

    def get_config(self) -> ServerConfig:
        return self._config

    def config(self, consumer: Callable[[ServerConfig], None]):
        """
        获取配置
        """
        consumer(self._config)
        return self


    def get_processor(self) -> Processor:
        return self._processor

    def listen(self, listener: Listener):
        """
        设置监听器
        """
        if listener is not None:
            self._listener = listener
        return self

    async def prestop(self):
        await self.prestop_do()

    async def stop(self):
        await self.stop_do()

    async def on_open(self, s: Session):
        self._sessions.add(s)
        await RunUtils.waitTry(self._listener.on_open(s))

    async def on_message(self, s: Session, m: Message):
        await RunUtils.waitTry(self._listener.on_message(s, m))

    async def on_reply(self, s: Session, m: Message):
        await RunUtils.waitTry(self._listener.on_reply(s, m))

    async def on_send(self, s: Session, m: Message):
        await RunUtils.waitTry(self._listener.on_send(s, m))

    async def on_close(self, s: Session):
        self._sessions.remove(s)
        await RunUtils.waitTry(self._listener.on_close(s))

    async def on_error(self, s: Session, e:Exception):
        await RunUtils.waitTry(self._listener.on_error(s, e))

    async def prestop_do(self):
        tmp = list(self._sessions)
        for s1 in tmp:
            if s1.is_valid():
                await s1.preclose()

    async def stop_do(self):
        tmp = list(self._sessions)
        for s1 in tmp:
            if s1.is_valid():
                await s1.close()

        self._sessions.clear()

File: javascript/socketd/src/socketd/transport/server/ServerBase.ts
import {ChannelAssistant} from "../core/ChannelAssistant";
import {Listener} from "../core/Listener";
import {Processor} from "../core/Processor";
import {ProcessorDefault} from "../core/impl/ProcessorDefault";
import {Session} from "../core/Session";
import {SimpleListener} from "../core/listener/SimpleListener";
import {ServerConfig} from "./ServerConfig";
import {IoConsumer} from "../core/Typealias";
import {Message} from "../core/Message";
import {RunUtils} from "../../utils/RunUtils";
import {Server} from "./Server";

/**
 * 服务端基类
 *
 * @author noear
 * @since 2.0
 */
export abstract class ServerBase<T extends ChannelAssistant<any>> implements Server,Listener {
    protected _processor: Processor = new ProcessorDefault();
    protected _sessions: Set<Session> = new Set<Session>();
    protected _listener: Listener = new SimpleListener();

    private _config: ServerConfig;
    private _assistant: T;
    protected _isStarted: boolean;

    constructor(config: ServerConfig, assistant: T) {
        this._config = config;
        this._assistant = assistant;
        this._processor.setListener(this);
    }

    /**
     * 获取通道助理
     */
    getAssistant(): T {
        return this._assistant;
    }

    abstract getTitle(): string ;

    /**
     * 获取配置
     */
    getConfig(): ServerConfig {
        return this._config;
    }

    /**
     * 配置
     */
    config(configHandler: IoConsumer<ServerConfig>): Server {
        if (configHandler) {
            configHandler(this._config);
        }
        return this;
    }


    /**
     * 获取处理器
     */
    getProcessor(): Processor {
        return this._processor;
    }


    /**
     * 设置监听器
     */
    listen(listener: Listener): Server {
        if (listener) {
            this._listener = listener;
        }
        return this;
    }

    abstract start(): Server;

    prestop() {
        this.prestopDo();
    }

    stop() {
        this.stopDo();
    }

    onOpen(s: Session) {
        this._sessions.add(s);
        this._listener.onOpen(s);
    }

    onMessage(s: Session, m: Message) {
        this._listener.onMessage(s, m);
    }

    onReply(s: Session, m: Message) {
        this._listener.onReply(s, m);
    }

    onSend(s: Session, m: Message) {
        this._listener.onSend(s, m);
    }

    onClose(s: Session) {
        this._sessions.delete(s);
        this._listener.onClose(s);
    }

    onError(s: Session, e: any) {
        this._listener.onError(s, e);
    }

    protected prestopDo() {
        for (let s1 of this._sessions) {
            if (s1.isValid()) {
                RunUtils.runAndTry(() => s1.preclose());
            }
        }
    }

    protected stopDo() {
        for (let s1 of this._sessions) {
            if (s1.isValid()) {
                RunUtils.runAndTry(() => s1.close());
            }
        }
        this._sessions.clear();
    }
}
